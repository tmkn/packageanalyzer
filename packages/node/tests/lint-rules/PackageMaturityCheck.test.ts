import { describe, test, expect, vi, beforeAll, afterAll } from "vitest";
import dayjs from "dayjs";

import {
    PackageMaturityCheck,
    type IPackageMaturityCheckParams
} from "../../src/reports/lint/checks/PackageMaturity.js";

import { createMockPackage } from "../../../test-utils/src/mocks.js";

describe(`PackageMaturityCheck`, () => {
    const now = dayjs(`2021-10-26`);

    beforeAll(() => {
        vi.setSystemTime(now.toDate());
    });

    afterAll(() => {
        vi.useRealTimers();
    });

    function scaffoldPackage(publicationTime: string) {
        return createMockPackage({
            name: "rolo",
            version: "1.0.0",
            attachments: {
                metafile: {
                    metaFile: {
                        time: {
                            "1.0.0": publicationTime
                        }
                    }
                }
            }
        });
    }

    test(`correctly returns message if package is younger than timespan`, () => {
        const now = dayjs();
        const publicationTime = now.subtract(1, "day");
        const pkg = scaffoldPackage(publicationTime.toISOString());

        const rule = new PackageMaturityCheck();
        const result = rule.check(pkg, { timespan: 2 });

        expect(result).toMatchSnapshot();
    });

    test(`should not return a message if package is mature (older than timespan)`, () => {
        const now = dayjs();
        const publicationTime = now.subtract(3, "day");
        const pkg = scaffoldPackage(publicationTime.toISOString());

        const rule = new PackageMaturityCheck();
        const result = rule.check(pkg, { timespan: 2 });

        expect(result).toBeUndefined();
    });

    test(`should not return a message if package is unpublished`, () => {
        const pkg = createMockPackage({
            name: "rolo",
            version: "1.0.0",
            attachments: {
                metafile: {
                    metaFile: {
                        time: {
                            unpublished: {}
                        }
                    }
                }
            }
        });

        const rule = new PackageMaturityCheck();
        const result = rule.check(pkg, { timespan: 2 });

        expect(result).toBeUndefined();
    });

    test(`should throw if version can't be found`, () => {
        const pkg = createMockPackage({
            name: "rolo",
            version: "1.0.0",
            attachments: {
                metafile: {
                    metaFile: {
                        time: {}
                    }
                }
            }
        });

        const rule = new PackageMaturityCheck();
        const check = () => rule.check(pkg, { timespan: 2 });

        expect(check).toThrowError(`Could not find publication time for version 1.0.0`);
    });

    describe(`checkParams`, () => {
        test(`correctly parses valid input`, () => {
            const rule = new PackageMaturityCheck();
            const schema = rule.checkParams();

            const validInput: IPackageMaturityCheckParams = { timespan: 5 };
            const validationResult = schema.safeParse(validInput);
            expect(validationResult.success).toBe(true);
        });

        test(`fails on invalid input`, () => {
            const rule = new PackageMaturityCheck();
            const schema = rule.checkParams();

            // @ts-expect-error testing invalid input
            const invalidInput: IPackageMaturityCheckParams = { timespan: "five" };
            const invalidResult = schema.safeParse(invalidInput);
            expect(invalidResult.success).toBe(false);
        });
    });

    describe(`pluralization`, () => {
        test(`correctly uses "days"`, () => {
            const now = dayjs();
            const publicationTime = now.subtract(2, "day");
            const pkg = scaffoldPackage(publicationTime.toISOString());

            const rule = new PackageMaturityCheck();
            const result = rule.check(pkg, { timespan: 4 });

            expect(result).toMatchSnapshot();
        });

        test(`correctly uses "day"`, () => {
            const now = dayjs();
            const publicationTime = now.subtract(1, "day");
            const pkg = scaffoldPackage(publicationTime.toISOString());

            const rule = new PackageMaturityCheck();
            const result = rule.check(pkg, { timespan: 2 });

            expect(result).toMatchSnapshot();
        });

        test(`correctly uses "day" & "days"`, () => {
            const now = dayjs();
            const publicationTime = now.subtract(1, "day");
            const pkg = scaffoldPackage(publicationTime.toISOString());

            const rule = new PackageMaturityCheck();
            const result = rule.check(pkg, { timespan: 3 });

            expect(result).toMatchSnapshot();
        });
    });
});
