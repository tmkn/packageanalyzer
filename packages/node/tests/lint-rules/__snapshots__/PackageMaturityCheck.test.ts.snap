// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PackageMaturityCheck > correctly returns message if package is younger than timespan 1`] = `"rolo@1.0.0 is 1 day old (published Oct 25, 2021, 12:00 AM); 1 more day required to reach 2-day maturity"`;

exports[`PackageMaturityCheck > pluralization > correctly uses "day" & "days" 1`] = `"rolo@1.0.0 is 1 day old (published Oct 25, 2021, 12:00 AM); 2 more days required to reach 3-day maturity"`;

exports[`PackageMaturityCheck > pluralization > correctly uses "day" 1`] = `"rolo@1.0.0 is 1 day old (published Oct 25, 2021, 12:00 AM); 1 more day required to reach 2-day maturity"`;

exports[`PackageMaturityCheck > pluralization > correctly uses "days" 1`] = `"rolo@1.0.0 is 2 days old (published Oct 24, 2021, 12:00 AM); 2 more days required to reach 4-day maturity"`;
